src/main/java/io/usethesource/vallang/impl/primitive/StringValue.java

```diff
@@ -1146,6 +1163,47 @@ public void write(Writer w) throws IOException {
             right.write(w);
         }
 
+        @Override
+        public Reader asReader() {
+            return new Reader() {
```

Comment:
Here we allocate a reader at every level in the tree. For the writer this design didn't hold up because of the enormous amounts of concat nodes in some applications. Typically we have as many concat nodes as lines in the output. 


So let's have a look at this and also use the benchmarks in the main method so we can fine-tune it. Typically a recursion is ok for the concat nodes and maximally log(n) allocations (a spine of reader objects that moves over the tree).

Reply:
The problem is that we have to keep state, the writer was easier as it's push, while the reader is pull. I tried to do it in a similar way as the writer, but could not in a reasonable timeframe figure it out. We need co-routintes ;)